carapax-i18n/src/lib.rs

```diff
@@ -0,0 +1,148 @@
+//! An i18n handler for carapax
+
+#![warn(missing_docs)]
+
+#[cfg(feature = "gettext")]
+mod gettext;
+#[cfg(feature = "gettext")]
+pub use gettext::*;
+
+#[cfg(feature = "fluent")]
+#[macro_use]
+extern crate rental;
+#[cfg(feature = "fluent")]
+mod fluent;
+#[cfg(feature = "fluent")]
+pub use fluent::*;
+
+use carapax::{core::types::Update, Context, HandlerFuture, HandlerResult, UpdateHandler};
+use std::collections::HashMap;
+
+/// An i18n handler for carapax
+#[derive(Debug, Clone)]
+pub struct I18nHandler<R, T> {
+    resolver: R,
+    default: T,
+    languages: Vec<T>,
+}
+
+impl<R, T> I18nHandler<R, T> {
+    /// Creates new I18nHandler
+    ///
+    /// Search translator in languages or use default one
+    pub fn new<I>(resolver: R, default: T, languages: I) -> Self
+    where
+        I: IntoIterator<Item = T>,
+    {
+        Self {
+            resolver,
+            default,
+            languages: languages.into_iter().collect(),
+        }
+    }
+}
+
+impl<R, T> UpdateHandler for I18nHandler<R, T>
+where
+    R: Resolver<T>,
+    T: Translator + Send + Sync + 'static,
+{
+    fn handle(&self, context: &mut Context, update: &Update) -> HandlerFuture {
+        let translator = self.resolver.resolve(update, &self.default, &self.languages);
+        context.set(translator);
+        HandlerResult::Continue.into()
+    }
+}
+
+/// An translator trait
+///
+/// Implement it to use your own i18n library
+pub trait Translator: Clone {
+    /// An error will be returned, if translation is failed
+    type Error;
+
+    /// Translates string with given arguments
+    fn translate(&self, id: &str, args: &Arguments) -> Result<String, Self::Error>;
+
+    /// Returns available locales
+    fn locales(&self) -> Vec<String>;
+}
+
+/// A resolver trait
+pub trait Resolver<T> {
+    /// Resolves [Translator](Translator)
+    fn resolve(&self, update: &Update, default: &T, languages: &[T]) -> T;
```

Comment:
I guess LocaleResolver should return a locale string from update and nothing else.